### ISSUE TITLE
[Android] System.ObjectDisposedException: Cannot access a disposed object. Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.

### ISSUE THREAD

**Issue by jassmith** [xamarin/xamarin-android#1520]
_From @borbmizzet on March 27, 2018 22:38_

### Description
My app was crashing intermittently with a "System.ObjectDisposedException".  I have no idea what was causing it, but it seems to have something to do with an IsEnabled Property binding.  I was never able to trigger it while debugging, so all I have is the crash report from Visual Studio app center.  Sorry for the lack of information.
### Steps to Reproduce

Unknown 

### Expected Behavior
Not crashing
### Actual Behavior
Crashing
### Basic Information

- Version with issue:
- Last known good version:
- IDE: Visual Studio 15.6.3
- Platform Target Frameworks: Android
  - Android: 8.1
- Android Support Library Version: 26.1.0.1
- Nuget Packages: Xamarin.Forms(.Pages) 2.5.0.280555, Acr.UserDialogs 7.0.1
- Affected Devices: Samsung Galaxy Note 5

### Screenshots

Android: 7.0
Android Build: NRD90M
Manufacturer: samsung
Model: SM-N920V
Start Date: 2018-03-26T19:05:24.761Z
Date: 2018-03-26T21:05:33.116Z

Xamarin Exception Stack:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <2a6afc2362644d56900ae1012edc48ef>:0
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <2a6afc2362644d56900ae1012edc48ef>:0
  at Android.Views.View.set_Enabled (System.Boolean value) [0x00022] in <12a8fd6f0b2d401b929358ab4a2a70f3>:0
  at Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateIsEnabled () [0x0001b] in <173f39d71f0d4d928f5bbea42e96ffa8>:0
  at Xamarin.Forms.Platform.Android.VisualElementTracker.HandlePropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x0035d] in <173f39d71f0d4d928f5bbea42e96ffa8>:0
  at (wrapper delegate-invoke) <Module>.invoke_void_object_PropertyChangedEventArgs(object,System.ComponentModel.PropertyChangedEventArgs)
  at Xamarin.Forms.BindableObject.OnPropertyChanged (System.String propertyName) [0x00012] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.Element.OnPropertyChanged (System.String propertyName) [0x00000] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.Internals.SetValueFlags attributes, System.Boolean silent) [0x0010e] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindableObject.SetValueCore (Xamarin.Forms.BindableProperty property, System.Object value, Xamarin.Forms.Internals.SetValueFlags attributes, Xamarin.Forms.BindableObject+SetValuePrivateFlags privateAttributes) [0x0015b] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindingExpression.ApplyCore (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property, System.Boolean fromTarget) [0x001f9] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindingExpression.Apply (System.Boolean fromTarget) [0x0003e] in <3754ae802c714b298131b9c985b370ad>:0
  at Xamarin.Forms.BindingExpression+BindingExpressionPart.<PropertyChanged>b__47_0 () [0x00000] in <3754ae802c714b298131b9c985b370ad>:0
  at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in <12a8fd6f0b2d401b929358ab4a2a70f3>:0
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <12a8fd6f0b2d401b929358ab4a2a70f3>:0
  at (wrapper dynamic-method) System.Object.3d44e075-a8ab-474b-93ee-c5de1f600909(intptr,intptr)

Thread 313:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Daemons$ReferenceQueueDaemon.run(Daemons.java:150)
2   java.lang.Thread.run(Thread.java:762)

Thread 314:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Object.wait(Object.java:407)
2   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:188)
3   java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:209)
4   java.lang.Daemons$FinalizerDaemon.run(Daemons.java:204)
5   java.lang.Thread.run(Thread.java:762)

Thread 315:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Daemons$FinalizerWatchdogDaemon.sleepUntilNeeded(Daemons.java:269)
2   java.lang.Daemons$FinalizerWatchdogDaemon.run(Daemons.java:249)
3   java.lang.Thread.run(Thread.java:762)

Thread 316:
0   dalvik.system.VMRuntime.runHeapTasks(VMRuntime.java:-2)
1   java.lang.Daemons$HeapTaskDaemon.run(Daemons.java:433)
2   java.lang.Thread.run(Thread.java:762)

Thread 322:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:323)
2   android.os.Looper.loop(Looper.java:136)
3   android.os.HandlerThread.run(HandlerThread.java:61)

Thread 324:
0   java.lang.Object.wait(Object.java:-2)
1   java.lang.Thread.parkFor$(Thread.java:2128)
2   sun.misc.Unsafe.park(Unsafe.java:325)
3   java.util.concurrent.locks.LockSupport.park(LockSupport.java:161)
4   java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2035)
5   java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:413)
6   java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1058)
7   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1118)
8   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
9   java.lang.Thread.run(Thread.java:762)

Thread 325:
0   android.os.MessageQueue.nativePollOnce(MessageQueue.java:-2)
1   android.os.MessageQueue.next(MessageQueue.java:323)
2   android.os.Looper.loop(Looper.java:136)
3   android.os.HandlerThread.run(HandlerThread.java:61)

Thread 338:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.okio.AsyncTimeout.awaitTimeout(AsyncTimeout.java:311)
2   com.android.okhttp.okio.AsyncTimeout.-wrap0(AsyncTimeout.java:-1)
3   com.android.okhttp.okio.AsyncTimeout$Watchdog.run(AsyncTimeout.java:286)

Thread 572:
0   java.lang.Object.wait(Object.java:-2)
1   com.android.okhttp.ConnectionPool.performCleanup(ConnectionPool.java:319)
2   com.android.okhttp.ConnectionPool.runCleanupUntilPoolIsEmpty(ConnectionPool.java:256)
3   com.android.okhttp.ConnectionPool.-wrap0(ConnectionPool.java:-1)
4   com.android.okhttp.ConnectionPool$1.run(ConnectionPool.java:102)
5   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
6   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
7   java.lang.Thread.run(Thread.java:762)

### Reproduction Link

_Copied from original issue: xamarin/Xamarin.Forms#2199_

**Comment by jassmith** [comment]
_From @heksan on March 28, 2018 8:52_

I can confirm this bug occurs, since I cannot paste my code in, here's overview of page it happens on:

ListView with  CachingStrategy="RecycleElement" and toolbar item changing appearance of the items(both inside and outside of the list) every time you tap it (using "IsVisible" and "IsEnabled" flags) . I am able to trigger this ONLY after i renavigate to this page multiple times.

Android 6.0
Android Support Library Version: 26.1.0.1
Nuget Packages: Xamarin.Forms 2.5.0.280555
Stack(in my case suggests it throws on UpdateIsVisible():

 2018-03-28 08:46:35.4201 ERROR MainActivity.AndroidEnvironment_UnhandledExceptionRaiser: Cannot access a disposed object.
 Object name: 'Xamarin.Forms.Platform.Android.Platform+DefaultRenderer'.
   at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <54816278eed9488eb28d3597fecd78f8>:0 
   at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualInt32Method (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <54816278eed9488eb28d3597fecd78f8>:0 
   at Android.Views.View.get_Visibility () [0x0000a] in <28e323a707a2414f8b493f6d4bb27c8d>:0 
   at Xamarin.Forms.Platform.Android.VisualElementTracker.UpdateIsVisible () [0x00017] in D:\agent\_work\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:295 
   at Xamarin.Forms.Platform.Android.VisualElementTracker.HandlePropertyChanged (System.Object sender, System.ComponentModel.PropertyChangedEventArgs e) [0x002e5] in D:\agent\_work\1\s\Xamarin.Forms.Platform.Android\VisualElementTracker.cs:157 
   at (wrapper delegate-invoke) <Module>:invoke_void_object_PropertyChangedEventArgs (object,System.ComponentModel.PropertyChangedEventArgs)
   at Xamarin.Forms.BindableObject.OnPropertyChanged (System.String propertyName) [0x0000a] in D:\agent\_work\1\s\Xamarin.Forms.Core\BindableObject.cs:138 2018-03-28 08:46:35.5249 DEBUG LocalMessageCacheRepository:CacheLocalMessage Successfully cached message [4136836a-2ac2-4f63-8295-b7e20abf4902]


   at Xamarin.Forms.Element.OnPropertyChanged (System.String propertyName) [0x00000] in D:\agent\_work\1\s\Xamarin.Forms.Core\Element.cs:401 
   at Xamarin.Forms.BindableObject.SetValueActual (Xamarin.Forms.BindableProperty property, Xamarin.Forms.BindableObject+BindablePropertyContext context, System.Object value, System.Boolean currentlyApplying, Xamarin.Forms.Internals.SetValueFlags attributes, System.Boolean silent) [0x00107] in D:\agent\_work\1\s\Xamarin.Forms.Core\BindableObject.cs:594 

   at Xamarin.Forms.BindingExpression.ApplyCore (System.Object sourceObject, Xamarin.Forms.BindableObject target, Xamarin.Forms.BindableProperty property, System.Boolean fromTarget) [0x001f9] in D:\agent\_work\1\s\Xamarin.Forms.Core\BindingExpression.cs:174 2018-03-28 08:46:35.5430 DEBUG AsyncCommunicationsController:QueueMessage QueueMessage: Start for 6c36cde9-8487-4f3e-83a4-6f5df4657bce

   at Xamarin.Forms.BindingExpression.Apply (System.Boolean fromTarget) [0x0003e] in D:\agent\_work\1\s\Xamarin.Forms.Core\BindingExpression.cs:56 
   at Xamarin.Forms.BindingExpression+BindingExpressionPart.<PropertyChanged>b__47_0 () [0x00000] in D:\agent\_work\1\s\Xamarin.Forms.Core\BindingExpression.cs:547 

   at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in <28e323a707a2414f8b493f6d4bb27c8d>:0 
   at (wrapper dynamic-method) System.Object:bb8cdb71-51a9-468c-b751-14fb3507f4ae (intptr,intptr)

**Comment by jassmith** [comment]
_From @PureWeen on March 28, 2018 20:36_

Is this a duplicate of
https://github.com/xamarin/Xamarin.Forms/issues/1846
?

**Comment by jassmith** [comment]
_From @heksan on March 29, 2018 12:26_

No, I mean underlying cause might be the same but in this bug there is no orientation change triggered in the code or manually. Moreover I have been trying to single out the reason for this crash by commenting out different parts of xaml but failed. Issue occurs intermittently when tapping the toolbar item.

**Comment by jassmith** [comment]
I moved the wrong issue, terribly sorry

**Comment by Kation** [comment]
I got this exception when I bind a model to a view.
I'm debugging **VisualElementTracker** and found that **HandlePropertyChanged** method called by **BindingExpression**.
In the method that **_renderer** is not null but **_renderer.View** was disposed.
And it will continue execute underlying codes.

Then I figure out that why it was happened.
I have a **thread running background** that receive messages from server will **changes model values**.
If a **page pop or push** while a message received and change the model values what is binding to old page's element, this exception will be throw.

I cannot find a way to detect is native view was disposed.
**_renderer.Element.EffectControlProvider** is also not null, means that **OnDetached of Effect** is not call at this moment.

Please reopen this issues.
